_posts/2023-10-23-unidirectional-topic-operator.md

```diff
@@ -0,0 +1,312 @@
+
+The first generation of the Topic Operator is based on the Kafka Streams library, and uses ZooKeeper to gather topic metadata.
+We call it **Bidirectional Topic Operator (BTO)**, as it allows to reconcile changes coming from both Kubernetes and Kafka using a three-way merge logic.
+In order to do this, it has to keep an internal topic store, which is the source-of-truth used to determine the necessary changes for synchronization.
+If the Kafka-side configuration is changed out-of-band (e.g. by using the Admin client, or Kafka shell scripts), those changes will eventually be reverted.
+
+The new generation of the Topic Operator is called **Unidirectional Topic Operator (UTO)**.
+The UTO improves on scalability and maintenance, but also provides support for Kafka clusters running in KRaft mode.
+It is fully backwards compatible with the BTO, but it will only reconcile topic events in one direction, from Kubernetes to Kafka.
+Your `KafkaTopic` custom resources will be the source-of-truth, like with all the other operators.
+
+The operator detects cases where multiple `KafkaTopic` resources are attempting to manage a Kafka topic using the same `.spec.topicName`.
+Only the oldest resource will be reconciled, while the other resource will fail with a `ResourceConflict` error.
+
+We still do not support decreasing `.spec.partitions` and changing `.spec.replicas`.
+While the former is not supported by Kafka itself, support for the latter could be added in the future.
+
+In addition to be able to pause reconciliations, the operator now also provides a way to unmanage a Kafka topic using annotations.
+If the `KafkaTopic` is paused and we delete it, then it will be also deleted in Kafka.
+If the `KafkaTopic` is unmanaged and we delete it, then only the resource will be garbage collected by Kubernetes, but the topic will still exists in Kafka.
+
+In terms of requirements, the UTO assumes the following grants:
+
+- describe all topics
+- describe all topics configs
+- create topics
+- create partitions
+- delete topics
+- list partition reassignments
+- describe broker configs
+
+#### Scalability
+
+The BTO is limited in terms of scalability, as it only operates on one topic at a time, and it has to update a persistent store containing topic metadata.
+Instead, the UTO does not store topic metadata and it aims to be scalable in terms of the number of topics that it can operate on.
+The following line graph confirms the BTO scalability issue, while the UTO scales almost linearly.
+
+<figure>
+    <img src="/assets/images/posts/2023-10-23-uto-max-recon-time.png" height=350>
+    <figcaption><small>
+        Fig 2. Line graph comparing BTO and UTO max reconciliation times.<br/>
+        We used a 3-nodes cluster with default configurations, running on a local Minikube instance with 10 cores and 28 GB of memory.<br/>
+        The test driver application was running on the same machine, and the time spent on the event queue was not included.
```

Comment:
This is still misleading. When you move from event-driven architecture to batch architecture the per-event efficiency almost always improves. But what matters here is the _time-to-reaction_. How long does it take from the event happening till the reconciliation is complete. So this needs to include the time it was waiting in the queue or be removed. I would expect it to come out as a better number even with the queue - at least when running it at scale.

Reply:
I'm comparing apples to apples here. We only measure reconciliation time, which does not include queuing time. 

The aim here is to show that the reconciliation is faster and more predictable, which is something the user can verify using provided metrics. 

In the other test mentioned below, I'm also reporting the readiness time which includes queuing time.

That said I'm open to proposals here.

Reply:
You might be comparing the same fruits. But it does not seem like apples but more like some fruit that nobody eats. What value does this number give me as a user? I don't care how long the reconciliation takes. I care about the time from the event happening in Kubernetes to the end of the reconciliation. As a user, I don't care that your reconciliation time is constant if the time in the queue is increasing to achieve that. This is important because you trade between more efficient reconciliation against time spent in queue. (The complete time is also what matters in the follow-up chapter where you talk about the race condition -> the window for it is defined by the time spent in queue and reconciliation.)

Reply:
Fair enough, I'm doing some more load tests and then will update that section.